jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java

```diff
@@ -457,6 +458,13 @@ else if (filled == 0)
         finally
         {
             setCurrentConnection(last);
+
+            // If handling has completed
+            if (_stream.get() == null && _retainableByteBuffer != null)
```

Comment:
This condition is tricky to grasp and gives a sense of fishiness at first sight, as it happened for @sbordet and myself too.

After carefully reading the explanations below, I believe this is correct although quite delicate so IMHO the following should be carefully stated in a comment:

 - when `_stream.get() == null`, there will never be another thread entering here
 - `while(getEndPoint().isOpen())` becoming an unreleasing noop if the endpoint is closed

Reply:
I'm going to defer the release fixes to another PR .